pkg/eval/builtin_fn_io.go

```diff
@@ -194,6 +195,180 @@ func print(fm *Frame, opts printOpts, args ...interface{}) {
+// to the byte stream.
+//
+// See Go's [`fmt`](https://golang.org/pkg/fmt/#hdr-Printing) package for
+// details about the formatting verbs and the various flags that modify the
+// default behavior; e.g., left versus right justification. We explicitly
+// handle only strings and float64 data types. Printing other types is likely
+// to result in unexpected output (but not an exception). You can use the
+// numeric formatting verbs (e.g., `%d`, `%f`) with either a float64 or a
+// string that can be converted to a number. If you use a float64 value to
+// the `%s` verb it will first be converted to a string using the rules for
+// [numbers](language.html#number]).
+//
+// Examples:
+//
+// ```elvish-transcript
+// ~> printf "%10s %.2f\n" Pi $math:pi
+//         Pi 3.14
+// ~> printf "%-10s %.2f %s\n" Pi $math:pi $math:pi
+// Pi         3.14 3.141592653589793
+// ~> printf "%d\n" 0b11100111
+// 231
+// ~> printf "%08b\n" 231
+// 11100111
+// ```
+//
+// **Note**: A couple of Go's [formatting
+// verbs](https://golang.org/pkg/fmt/#hdr-Printing), such as `%b` and `%x`,
+// behave differently based on the associated value type. In Elvish the
+// type is always an integer; specifically, a Go `int`; regardless of whether
+// the Elvish value is a string or float64. That is, the number is coerced to
+// an int if possible else an exception is raised.
+//
+// **Note**: Do not use the `%p` or `%q` formatting verbs at this time. The `%p`
+// verb will never have any meaning. The `%q` formatting verb will eventually
+// behave as if you had printed the result of `repr` on the value. Similarly,
+// only scalars such as strings and numbers are currently supported. Passing
+// an Elvish list, map, exception, or other complex type will produce
+// unexpected output. This will be improved in the future.
+//
+// **Note**: This is loosely based on the [POSIX `printf`
+// command](https://pubs.opengroup.org/onlinepubs/007908799/xcu/printf.html).
+// There are three notable differences. The first is that the formatting verbs
+// and behavior use Go's [`fmt`](https://golang.org/pkg/fmt/) package and not
+// the POSIX specification. The two have many similarities but are not
+// identical. The second is that the number of values must match the number of
+// formatting verbs. Excess values do not result in the format being evaluated
+// additional times until all values are consumed as happens with the POSIX
+// command of the same name. You must explicitly split the value list and
+// invoke this builtin for each block of values. The third is that the Elvish
+// `printf` does not recognize backslash sequences such as `\n`. You must use
+// [double-quoted Elvish strings](language.html#double-quoted-string) if you
+// want such backslash sequences to be recognized -- just as you would with a
+// string passed to any other Elvish builtin.
+//
+// @cf print echo pprint repr
+
+type formatterString struct {
+	str string
+}
+type formatterFloat64 struct {
+	num float64
+}
```

Comment:
Once you think of the verbs as descriptions of how to *convert* the argument, it should become clear that it's not necessary to define different formatter for different types.

The branch for `%s` and `%v` can both use `vals.ToString`, the branch for `%d`, `%f` and other numerical verbs can all use `vals.ScanToGo` with a `int` or `float64` destination.